### PR TITLE
ddl, expression: Disallow add stored generated columns through `ALTER TABLE`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve? 

According to current [TiDB document](https://github.com/pingcap/docs/blob/master/v2.1/sql/generated-columns.md#limitations), adding stored generated columns are not supported:

```
You cannot add the generated column in the storage type of STORED through ALTER TABLE.
```

So, an error should be returned for such statements.

### What is changed and how it works?

- A check in AddColumn()
- Other redability improvements

Check the properties of the new column, if it is both `Generated` and `Stored`, return an `errUnsupportedOnGeneratedColumn` error.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: ddl/generated_column.go [hunk 1]
```diff
@@ -195,6 +195,13 @@ func checkIllegalFn4GeneratedColumn(colName string, expr ast.ExprNode) error {
 	return nil
 }
 
+func checkAddingGeneratedStoredColumn(option *ast.ColumnOption) error {
+	if option.Tp == ast.ColumnOptionGenerated && option.Stored {
+		return errUnsupportedOnGeneratedColumn.GenWithStackByArgs("Adding STORED column through `alter table`")
```
Comment (crazycs520) [inline ddl/generated_column.go]: s/STORED/generated stored?

File: ddl/generated_column.go [hunk 2]
```diff
@@ -195,6 +195,13 @@ func checkIllegalFn4GeneratedColumn(colName string, expr ast.ExprNode) error {
 	return nil
 }
 
+func checkAddingGeneratedStoredColumn(option *ast.ColumnOption) error {
+	if option.Tp == ast.ColumnOptionGenerated && option.Stored {
```
Comment (crazycs520) [inline ddl/generated_column.go]: `option.Tp == ast.ColumnOptionGenerated` already checked in line#1992.

File: ddl/db_test.go [hunk 3]
```diff
@@ -2020,23 +2020,18 @@ func (s *testDBSuite3) TestGeneratedColumnDDL(c *C) {
 	s.tk.MustExec("use test")
 
 	// Check create table with virtual generated column.
```
Comment (bb7133) [inline ddl/db_test.go]:
```suggestion
	// Check create table with virtual and stored generated columns.
```
Comment should be updated as well.

File: ddl/db_test.go [hunk 4]
```diff
@@ -2020,23 +2020,18 @@ func (s *testDBSuite3) TestGeneratedColumnDDL(c *C) {
...
-	s.tk.MustExec(`CREATE TABLE test_gv_ddl(a int, b int as (a+8) virtual)`)
+	s.tk.MustExec(`CREATE TABLE test_gv_ddl(a int, b int as (a+8) virtual, c int as (b + 2) stored)`)
 
 	// Check desc table with virtual generated column.
```
Comment (bb7133) [inline ddl/db_test.go]:
```suggestion
	// Check desc table with virtual and stored generated columns.
```
Ditto

File: ddl/db_test.go [hunk 5]
```diff
@@ -2020,23 +2020,18 @@ func (s *testDBSuite3) TestGeneratedColumnDDL(c *C) {
...
 	result := s.tk.MustQuery(`DESC test_gv_ddl`)
-	result.Check(testkit.Rows(`a int(11) YES  <nil> `, `b int(11) YES  <nil> VIRTUAL GENERATED`))
+	result.Check(testkit.Rows(`a int(11) YES  <nil> `, `b int(11) YES  <nil> VIRTUAL GENERATED`, `c int(11) YES  <nil> STORED GENERATED`))
 
 	// Check show create table with virtual generated column.
```
Comment (bb7133) [inline ddl/db_test.go]:
```suggestion
	// Check show create table with virtual and stored generated columns.
```
Ditto

File: ddl/ddl.go [hunk 6]
```diff
@@ -103,13 +103,14 @@ var (
...
-	errErrorOnRename            = terror.ClassDDL.New(codeErrorOnRename, "Error on rename of './%s/%s' to './%s/%s'")
-	errBadField                 = terror.ClassDDL.New(codeBadField, "Unknown column '%s' in '%s'")
+	errKeyColumnDoesNotExits = terror.ClassDDL.New(codeKeyColumnDoesNotExits, mysql.MySQLErrName[mysql.ErrKeyColumnDoesNotExits])
+	errUnknownTypeLength     = terror.ClassDDL.New(codeUnknownTypeLength, "Unknown length for type tp %d")
+	errUnknownFractionLength = terror.ClassDDL.New(codeUnknownFractionLength, "Unknown Length for type tp %d and fraction %d")
+	errInvalidJobVersion     = terror.ClassDDL.New(codeInvalidJobVersion, "DDL job with version %d greater than current %d")
+	errFileNotFound          = terror.ClassDDL.New(codeFileNotFound, "Can't find file: './%s/%s.frm'")
+	errErrorOnRename         = terror.ClassDDL.New(codeErrorOnRename, "Error on rename of './%s/%s' to './%s/%s'")
+	// ErrBadField forbids to refer to unknown column.
+	ErrBadField                 = terror.ClassDDL.New(codeBadField, "Unknown column '%s' in '%s'")
```
Comment (bb7133) [inline ddl/ddl.go]: Once being made public, `ErrBadField` should be moved to where other public errors are.

File: ddl/generated_column.go [hunk 7]
```diff
@@ -41,21 +41,24 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
...
-// columnNamesCover checks whether dependColNames is covered by normalColNames or not.
-// it's only for alter table add column because before alter, we can make sure that all
-// columns in table are verified already.
-func columnNamesCover(normalColNames map[string]struct{}, dependColNames map[string]struct{}) error {
-	for name := range dependColNames {
-		if _, ok := normalColNames[name]; !ok {
-			return errBadField.GenWithStackByArgs(name, "generated column function")
+// ensure all depended columns are exist.
+//
+// NOTE: this will destroy parameter `dependCols`
```
Comment (bb7133) [inline ddl/generated_column.go]: IMHO it's better to replace term 'destroy' with 'modify'... 'destroy' a map is confusing

File: ddl/generated_column.go [hunk 8]
```diff
@@ -41,21 +41,24 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
...
+func checkDependedColExist(dependCols map[string]struct{}, cols []*table.Column) error {
+	for _, col := range cols {
+		delete(dependCols, col.Name.L)
+	}
+	if len(dependCols) != 0 {
+		for arbitraryKey := range dependCols {
```
Comment (bb7133) [inline ddl/generated_column.go]: the `for` loop is not needed, just using `dependCols[0]` is enough

Comment (tangenta) [reply ddl/generated_column.go]: `dependCols` is a map, maybe `for` loop is the only choice

File: ddl/generated_column.go [hunk 9]
```diff
@@ -41,21 +41,27 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
...
-// it's only for alter table add column because before alter, we can make sure that all
-// columns in table are verified already.
-func columnNamesCover(normalColNames map[string]struct{}, dependColNames map[string]struct{}) error {
-	for name := range dependColNames {
-		if _, ok := normalColNames[name]; !ok {
-			return errBadField.GenWithStackByArgs(name, "generated column function")
+// ensure all depended columns are exist.
+func checkDependedColExist(dependCols map[string]struct{}, cols []*table.Column) error {
+	dependColsCopy := make(map[string]struct{}, len(dependCols))
+	for k, v := range dependCols {
```
Comment (crazycs520) [inline ddl/generated_column.go]: no need copy here.

File: ddl/generated_column.go [hunk 10]
```diff
@@ -41,21 +41,24 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
...
 }
 
-// columnNamesCover checks whether dependColNames is covered by normalColNames or not.
-// it's only for alter table add column because before alter, we can make sure that all
-// columns in table are verified already.
-func columnNamesCover(normalColNames map[string]struct{}, dependColNames map[string]struct{}) error {
-	for name := range dependColNames {
-		if _, ok := normalColNames[name]; !ok {
-			return errBadField.GenWithStackByArgs(name, "generated column function")
+// ensure all depended columns are exist.
```
Comment (bb7133) [inline ddl/generated_column.go]:
```suggestion
// ensure all depended columns exist.
```

File: ddl/generated_column.go (hunk 8 above)
Comment (bb7133) [reply ddl/generated_column.go]: Sure, forget that.

File: ddl/generated_column.go (hunk 10 above)
Comment (zimulala) [reply ddl/generated_column.go]: Function comments need to start with "functionName "

File: ddl/generated_column.go [hunk 11]
```diff
@@ -41,21 +41,24 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
...
+//
+// NOTE: this will MODIFY parameter `dependCols`
```
Comment (zimulala) [inline ddl/generated_column.go]: Please add `.` at the end of a line

File: ddl/db_test.go [hunk 12]
```diff
@@ -2071,6 +2066,10 @@ func (s *testDBSuite3) TestGeneratedColumnDDL(c *C) {
 		{`create table test_gv_ddl_bad (a int, b int, c int as (a+b) primary key)`, mysql.ErrUnsupportedOnGeneratedColumn},
 		{`create table test_gv_ddl_bad (a int, b int, c int as (a+b), primary key(c))`, mysql.ErrUnsupportedOnGeneratedColumn},
 		{`create table test_gv_ddl_bad (a int, b int, c int as (a+b), primary key(a, c))`, mysql.ErrUnsupportedOnGeneratedColumn},
+
+		// add stored generated column through alter table
```
Comment (zimulala) [inline ddl/db_test.go]:
```suggestion
		// Add stored generated column through alter table.

Comment (tangenta) [reply ddl/db_test.go]: All of the similar comments in this function are uncapitalized. Do I need to change them together?

Comment (zimulala) [reply ddl/db_test.go]: You can only change the comment in this PR first.

Comment (tangenta) [reply ddl/db_test.go]: Well, I change all of them for consistency. PTAL

**Pull request merged**